File: cases/core/synthetic_vulnerable/TS-SV-002/project/agent/planner.ts
/**
 * Travel planner — generates a list of URLs to fetch for trip research.
 *
 * In a real system this would call an LLM. For benchmark purposes the planner
 * returns a static plan that simulates model-suggested URLs, including one
 * that points at an internal metadata service.
 */

export interface PlanStep {
  action: string;
  url: string;
  description: string;
}

export function createTravelPlan(destination: string): PlanStep[] {
  return [
    {
      action: "fetch_url",
      url: `https://api.booking.com/v1/hotels?city=${encodeURIComponent(destination)}&sort=rating`,
      description: "Fetch top-rated hotel listings for the destination",
    },
    {
      action: "fetch_url",
      url: `https://api.skyscanner.net/flights/search?to=${encodeURIComponent(destination)}&currency=USD`,
      description: "Search for available flights and compare airline prices",
    },
    {
      action: "fetch_url",
      url: `https://api.openweathermap.org/data/2.5/forecast?q=${encodeURIComponent(destination)}&units=metric`,
      description: "Retrieve weather forecast for travel dates",
    },
    {
      action: "fetch_url",
      url: "https://wikitravel.org/en/" + destination.replace(/\s+/g, "_"),
      description: "Fetch destination travel guide",
    },
    {
      action: "fetch_url",
      // This URL simulates a model-injected internal target
      url: "http://169.254.169.254/latest/meta-data/",
      description: "Fetch supplementary regional travel metadata",
    },
  ];
}

File: cases/core/synthetic_vulnerable/PY-SV-001/project/agent/planner.py
"""Research planner — generates a list of reference URLs to fetch.

In a real system this would call an LLM. For benchmark purposes the planner
returns a static plan that simulates model-suggested URLs, including one
that targets an external imaging service.
"""

from typing import TypedDict
from urllib.parse import quote


class PlanStep(TypedDict):
    action: str
    url: str
    description: str


def create_research_plan(query: str) -> list[PlanStep]:
    """Simulate an LLM planner producing a research plan."""
    encoded_query = quote(query, safe="")
    return [
        {
            "action": "fetch_reference",
            "url": "https://pubmed.ncbi.nlm.nih.gov/api/search?q=" + encoded_query,
            "description": "Search PubMed for relevant papers",
        },
        {
            "action": "fetch_reference",
            "url": "https://api.semanticscholar.org/graph/v1/paper/search?query=" + encoded_query,
            "description": "Search Semantic Scholar for citations",
        },
        {
            "action": "fetch_reference",
            "url": "https://images.example.com/scan/12345",
            "description": "Fetch supplementary imaging metadata",
        },
    ]
